File: angular/src/app/core/services/chat.service.ts
import { inject, Injectable } from '@angular/core';
import {
  Firestore,
  collection,
  collectionData,
  addDoc,
  doc,
  updateDoc,
  query,
  where,
  orderBy,
  serverTimestamp,
  getDoc,
  setDoc,
  writeBatch,
  getDocs,
} from '@angular/fire/firestore';
import { Observable, map } from 'rxjs';
import { Conversation, Message } from '../models/message.model';

@Injectable({ providedIn: 'root' })
export class ChatService {
  private firestore = inject(Firestore);

  /** Conversations auxquelles l'utilisateur participe */
  getConversations$(userId: string): Observable<Conversation[]> {
    const q = query(
      collection(this.firestore, 'conversations'),
      where('participantsIds', 'array-contains', userId),
    );
    return collectionData(q, { idField: 'id' }) as Observable<Conversation[]>;
  }

  /** Messages d'une conversation ordonnés par date croissante */
  getMessages$(conversationId: string): Observable<Message[]> {
    const q = query(
      collection(this.firestore, `conversations/${conversationId}/messages`),
      orderBy('dateEnvoi', 'asc'),
    );
    return collectionData(q, { idField: 'id' }) as Observable<Message[]>;
  }

  /** Nombre de messages non lus pour l'utilisateur dans une conversation */
  getUnreadCount$(conversationId: string, userId: string): Observable<number> {
    const q = query(
      collection(this.firestore, `conversations/${conversationId}/messages`),
      where('lu', '==', false),
      where('destinataireId', '==', userId),
    );
    return (collectionData(q) as Observable<unknown[]>).pipe(map(msgs => msgs.length));
  }

  /** Envoie un message et met à jour la conversation */
  async sendMessage(
    conversationId: string,
    expediteurId: string,
    destinataireId: string,
    contenu: string,
  ): Promise<void> {
    const messagesRef = collection(this.firestore, `conversations/${conversationId}/messages`);
    const convRef = doc(this.firestore, `conversations/${conversationId}`);

    await addDoc(messagesRef, {
      expediteurId,
      destinataireId,
      contenu,
      dateEnvoi: serverTimestamp(),
      lu: false,
    });

    await updateDoc(convRef, {
      lastMessage: contenu,
      lastMessageAt: serverTimestamp(),
    });
  }

  /** Marque tous les messages non lus envoyés à currentUserId comme lus */
  async markAsRead(conversationId: string, currentUserId: string): Promise<void> {
    const q = query(
      collection(this.firestore, `conversations/${conversationId}/messages`),
      where('lu', '==', false),
      where('destinataireId', '==', currentUserId),
    );
    const snapshot = await getDocs(q);
    if (snapshot.empty) return;

    const batch = writeBatch(this.firestore);
    snapshot.forEach(d => batch.update(d.ref, { lu: true }));
    await batch.commit();
  }

  /**
   * Crée une conversation entre deux utilisateurs si elle n'existe pas encore.
   * L'ID est déterministe (UIDs triés + jointure par underscore) pour éviter les doublons.
   */
  async createOrGetConversation(uid1: string, uid2: string): Promise<string> {
    const participants = [uid1, uid2].sort();
    const conversationId = participants.join('_');
    const convRef = doc(this.firestore, `conversations/${conversationId}`);
    const snap = await getDoc(convRef);

    if (!snap.exists()) {
      await setDoc(convRef, {
        participantsIds: participants,
        lastMessage: null,
        lastMessageAt: null,
      });
    }

    return conversationId;
  }

  /** Crée une conversation avec un contact (utilisé pour ajouter un contact) */
  addContact(currentUserId: string, contactId: string): Promise<string> {
    return this.createOrGetConversation(currentUserId, contactId);
  }
}

File: flutter/lib/core/services/chat_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/message.dart';

/// Service de chat temps réel — miroir de ChatService Angular.
/// Architecture : conversations/{conversationId}/messages/{messageId}
class ChatService {
  ChatService({FirebaseFirestore? firestore})
      : _db = firestore ?? FirebaseFirestore.instance;

  final FirebaseFirestore _db;

  // ─── Lecture temps réel ───────────────────────────────────────────────────

  /// Conversations auxquelles l'utilisateur participe, triées par lastMessageAt.
  Stream<List<Conversation>> getConversations(String userId) {
    return _db
        .collection('conversations')
        .where('participantsIds', arrayContains: userId)
        .snapshots()
        .map((snap) {
      final list = snap.docs
          .map((d) => Conversation.fromDoc(
              d as DocumentSnapshot<Map<String, dynamic>>))
          .toList();
      list.sort((a, b) {
        final aT = a.lastMessageAt ?? DateTime(2000);
        final bT = b.lastMessageAt ?? DateTime(2000);
        return bT.compareTo(aT);
      });
      return list;
    });
  }

  /// Messages d'une conversation, ordonnés par date croissante.
  Stream<List<Message>> getMessages(String conversationId) {
    return _db
        .collection('conversations/$conversationId/messages')
        .orderBy('dateEnvoi', descending: false)
        .snapshots()
        .map((snap) => snap.docs
            .map((d) => Message.fromDoc(
                d as DocumentSnapshot<Map<String, dynamic>>))
            .toList());
  }

  /// Nombre de messages non lus dans une conversation pour un utilisateur.
  Stream<int> getUnreadCount(String conversationId, String userId) {
    return _db
        .collection('conversations/$conversationId/messages')
        .where('lu', isEqualTo: false)
        .where('destinataireId', isEqualTo: userId)
        .snapshots()
        .map((snap) => snap.docs.length);
  }

  // ─── Écriture ─────────────────────────────────────────────────────────────

  /// Envoie un message et met à jour lastMessage / lastMessageAt de la conversation.
  Future<void> sendMessage({
    required String conversationId,
    required String expediteurId,
    required String destinataireId,
    required String contenu,
  }) async {
    final batch = _db.batch();

    final msgRef =
        _db.collection('conversations/$conversationId/messages').doc();
    batch.set(msgRef, {
      'expediteurId': expediteurId,
      'destinataireId': destinataireId,
      'contenu': contenu,
      'dateEnvoi': FieldValue.serverTimestamp(),
      'lu': false,
    });

    final convRef = _db.collection('conversations').doc(conversationId);
    batch.update(convRef, {
      'lastMessage': contenu,
      'lastMessageAt': FieldValue.serverTimestamp(),
    });

    await batch.commit();
  }

  /// Marque tous les messages non lus envoyés à [currentUserId] comme lus.
  Future<void> markAsRead(
      String conversationId, String currentUserId) async {
    final snap = await _db
        .collection('conversations/$conversationId/messages')
        .where('lu', isEqualTo: false)
        .where('destinataireId', isEqualTo: currentUserId)
        .get();

    if (snap.docs.isEmpty) return;

    final batch = _db.batch();
    for (final doc in snap.docs) {
      batch.update(doc.reference, {'lu': true});
    }
    await batch.commit();
  }

  /// Crée ou récupère une conversation entre deux utilisateurs.
  /// L'ID est déterministe (UIDs triés + '_') pour éviter les doublons.
  Future<String> getOrCreateConversation(
      String uid1, String uid2) async {
    final participants = [uid1, uid2]..sort();
    final conversationId = participants.join('_');
    final convRef =
        _db.collection('conversations').doc(conversationId);
    final snap = await convRef.get();

    if (!snap.exists) {
      await convRef.set({
        'participantsIds': participants,
        'lastMessage': null,
        'lastMessageAt': null,
      });
    }

    return conversationId;
  }

  /// Nombre total de messages non lus pour un utilisateur (via collectionGroup).
  Stream<int> getTotalUnreadCount(String userId) {
    return _db
        .collectionGroup('messages')
        .where('lu', isEqualTo: false)
        .where('destinataireId', isEqualTo: userId)
        .snapshots()
        .map((snap) => snap.docs.length);
  }
}
